qutip/operators.py

```diff
@@ -422,20 +424,78 @@ def create(N, offset=0):
     return qo.dag()
 
 
+def _implicit_tensor_dimensions(dimensions):
+    """
+    Total flattened size and operator dimensions for operator creation routines
+    that automatically perform tensor products.
+
+    Parameters
+    ----------
+    dimensions : (int) or (list of int) or (list of list of int)
+        First dimension of an operator which can create an implicit tensor
+        product.  If the type is `int`, it is promoted first to `[dimensions]`.
+        From there, it should be one of the two-elements `dims` parameter of a
+        `qutip.Qobj` representing an `oper` or `super`, with possible tensor
+        products.
+
+    Returns
+    -------
+    size : int
+        Dimension of backing matrix required to represent operator.
+    dimensions : list
+        Dimension list in the form required by ``Qobj`` creation.
+    """
+    if not isinstance(dimensions, list):
+        dimensions = [dimensions]
+    flat = flatten(dimensions)
+    if not all(isinstance(x, numbers.Integral) and x >= 0 for x in flat):
+        raise ValueError("All dimensions must be integers >= 0")
+    dimensions = [dimensions, dimensions]
+    if type_from_dims(dimensions) not in ['oper', 'super']:
+        raise TypeError("`dimensions` must be a valid single dimension of a "
```

Comment:
`type_from_dims([[1],[1]])` is `"bra"`, not `"oper"`. So `qeye(1)` fails.

Reply:
This is a good point.  After a bit more testing, I think the handling of 1D Hilbert spaces is a rather inconsistent throughout `qutip`, with the bug here manifesting itself in a few different places.

Some things I found:
  1. `qutip.create(1).type == qutip.basis(1, 0).type == 'bra'`.  This is the same type-inference bug as you've pointed out here, but those two functions don't test for it, so it silently gives an unexpected result (there are probably more, as `create` is one I haven't touched, and the bug existed in the old `basis` too), which fortunately is then mostly ignored by other calling code.
  2. `1 + qutip.tensor(qutip.create(1), qutip.create(1))` raises `TypeError` for incompatible dimensions, but making either call `create(2)` succeeds (as it should).  Incidentally, `1 + qutip.basis(5, 0)` succeeds - adding a scalar to a ket -, which I personally find rather odd.
  3. Take a tensor-product operator where any one or more of the Hilbert spaces have one subspace of dimension one, for example `a = qutip.tensor(qutip.create(2), qutip.create(1))`.  Now `a + a` and `1 * a` are the correct dimensions, but `a * a` contracts all scalar dimensions to a maximum of one (and that remains only if _all_ the spaces' dimensions are one), so `a * a * a` is a `TypeError`.

Looking at the code, this final multiplication one seems to be deliberate, but it's pretty surprising behaviour in this case, no matter how mathematically useless those operators are, and it's probably a bug that addition and multiplication are inconsistent.  I assume this is to allow tensor-product bra/ket multiplication to contract to a canonical scalar representation.

### Possible fixes

`type_from_dims` could return a `set` of valid types rather than picking one at random, and allow the caller to decide what is sensible in this situation.  In this case, `Qobj` initialisation should fail loudly if the type isn't explicitly passed and there's more than one possibility. Alternatively, and also not entirely satisfactorily, the default for `type_from_dims([[1], [1]])` (or other equivalent scalar forms) could be `'oper'`, not `'bra'`.  It's not really either (could be a new `'scalar'` type?), but numeric types behave like they're multiplied by an implicit identity operator, so `'oper'` seems more reasonable.

Mostly, I think points 2 and 3 are inconsistent with each other, rather than either being necessarily an error.  Still, I should be able to add a scalar to a square operator of any tensor-product size, and the default behaviour allowing a square operator to have `a**3` be valid and `a*a*a` be a `TypeError` is wrong.  Perhaps there should be a check in the automatic dimension contraction to contract only if that dimension changed size compared to either `Qobj`?

Reply:
The questions is mostly how much energy do we want to spend on "scalar" Hilbert space, I don't see any use for them other than quick tests. I would be fine with simply removing the dimension check in `qeye`. It is redundant as a test.

`type_from_dims` returning a set of all possible type could be a good option. But I am not sure it's needed. Setting 'oper' as the default is probably sufficient. Not knowing the type should not cause an error.

`1 + qutip.tensor(qutip.create(1), qutip.create(1))` There is actually an exception to stop these cases. `1 + qutip.basis(5, 0)`: Quite surprising behavior. I expect this is an error that come from cut/pasting the product code. Updating `Qobj.__add__` deserve it's own issue.

3 - Setting `qutip.settings.auto_tidyup_dims = False` make `a*a*a` behave as expected. There is also an old PR to correct this: #735.

Reply:
In theory I don't like removing the test because I think library functions should be conservative in what they return, and it shouldn't be possible to use `qutip` library functions to get an invalid `Qobj`.  That said, `type_from_dims` assumes that the dimensions are properly formatted, so it wasn't a great test.

I've removed the test for now, but I think strengthening `Qobj.type` and the dimension checks could make it easier to handle some of the edge cases we were discussing above.

Reply:
For the other points:

If `def op(n): return qutip.tensor(qutip.create(n), qutip.create(n))`, then surely either `1 + op(1)` and `1 + op(2)` should both error, or neither of them should?  The exception at the moment catches it at the moment presumably to prevent things like `qutip.create(1) + qutip.tensor(qutip.create(1), qutip.create(1))`, which _would_ be an error, but because scalars are promoted to `Qobj` at the top of the function, they get caught up in it too.

`1 + qutip.basis(5, 0)` seems to have been added deliberately - there's an `if` branch to catch the case of `Qobj.type not in ['oper', 'super']`.  Without knowing the reason, it seems a very odd choice, especially because it only adds the scalar to the filled entries, which again, was clearly a deliberate decision in the code.

I'll have a look at the interplay of `Qobj.type`, `__add__` and `__mul__` and see if I can iron out some of those bugs without introducing a whole bunch more.  But that would be a different PR.